KAFKA-16223: Replace EasyMock/PowerMock with Mockito for KafkaConfigBackingStoreTest

chiacyu opened the pull request:
This pr cleans the powermock and easymock in KafkaConfigBackingStoreTest and I also write the corresponding tests with Mockito. Thanks.

chiacyu committed:
Initialized the commit

chiacyu committed:
Initialized the commit

chia7712 commented:
@chiacyu Could you please fix the conflicts?

chiacyu committed:
Reorder the tests

cmccabe committed:
KAFKA-16515: Fix the ZK Metadata cache confusion between brokers and controllers

ZkMetadataCache could theoretically return KRaft controller information from a call to
ZkMetadataCache.getAliveBrokerNode, which doesn't make sense. KRaft controllers are not part of the
set of brokers. The only use-case for this functionality was in MetadataCacheControllerNodeProvider
during ZK migration, where it allowed ZK brokers in migration mode to forward requests to
kcontrollers when appropriate. This PR changes MetadataCacheControllerNodeProvider to simply
delegate to quorumControllerNodeProvider in this case.

Reviewers: José Armando García Sancio <jsancio@apache.org>

cmccabe committed:
KAFKA-16516: Fix the controller node provider for broker to control channel

Fix the code in the RaftControllerNodeProvider to query RaftManager to find Node information,
rather than consulting a static map. Add a RaftManager.voterNode function to supply this
information. In KRaftClusterTest, add testControllerFailover to get more coverage of controller
failovers.

Reviewers: José Armando García Sancio <jsancio@apache.org>

chiacyu committed:
Reorder the unit tests

chiacyu committed:
Modify for solved merge conflict

chiacyu committed:
Merge trunk to solve conflict

chiacyu committed:
Removed old tests

lucasbru committed:
MINOR: fix streams_broker_compatibility test (#16015)

The system test was broken in ccf4bd5
which failed to import the matrix symbol. The test was failing silently, not
discovering any tests.

Reviewers: Bruno Cadonna <cadonna@apache.org>

TaiJuWu committed:
KAFKA-16654 Refactor kafka.test.annotation.Type and ClusterTestExtensions (#15916)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

FrankYang0529 committed:
KAFKA-16784 Migrate TopicBasedRemoteLogMetadataManagerMultipleSubscriptionsTest to use ClusterTestExtensions (#15992)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mimaison committed:
KAFKA-7632: Support Compression Levels (KIP-390) (#15516)


Reviewers: Jun Rao <jun@confluent.io>,  Luke Chen <showuon@gmail.com>
Co-authored-by: Lee Dongjin <dongjin@apache.org>

AyoubOm committed:
KAFKA-16343: Add unit tests of foreignKeyJoin classes (#15564)

Added unit tests of two processors included in foreignKey join : SubscriptionSendProcessorSupplier and ForeignTableJoinProcessorSupplier.
Renamed ForeignTableJoinProcessorSupplierTest to SubscriptionJoinProcessorSupplierTest as that's the processor which the test class is testing.

Reviewers: Walker Carlson <wcarlson@apache.org>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 6) Post process new assignment structure (#16002)

This PR creates the required methods to post-process the result of TaskAssignor.assign into the required ClientMetadata map. This allows most of the internal logic to remain intact after the user's assignment code runs.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

gharris1727 committed:
KAFKA-16739: Exclude protected members from aggregated release javadocs (#15940)

Reviewers: Matthias J. Sax <matthias@confluent.io>

FrankYang0529 committed:
KAFKA-16783: Migrate RemoteLogMetadataManagerTest to new test infra (#15983)

1. Replace TopicBasedRemoteLogMetadataManagerWrapperWithHarness with RemoteLogMetadataManagerTestUtils#builder in RemoteLogMetadataManagerTest.
2. Use ClusterTestExtention for RemoteLogMetadataManagerTest.

Signed-off-by: PoAn Yang <payang@apache.org>

Reviewers: Luke Chen <showuon@gmail.com>

kagarwal06 committed:
KAFKA-15444: Native docker image for Apache Kafka (KIP-974)  (#15927)

This PR aims to add Docker Image for GraalVM based Native Kafka Broker as per the following KIP - https://cwiki.apache.org/confluence/display/KAFKA/KIP-974%3A+Docker+Image+for+GraalVM+based+Native+Kafka+Broker

This PR adds the following functionalities:

Ability to build the docker image for Native Apache Kafka
   - Dockerfile
   - Launch script
   - metadata configs required by graalVM native-image: link
Add Kafka startup ability in the KafkaDockerWrapper.scala
Ability to build and test the image - integrated with the existing JVM docker image framework.

nicktelford committed:
KAFKA-15541: Add num-open-iterators metric (#15975)

Part of [KIP-989](https://cwiki.apache.org/confluence/x/9KCzDw).

This new `StateStore` metric tracks the number of `Iterator` instances
that have been created, but not yet closed (via `AutoCloseable#close()`).

This will aid users in detecting leaked iterators, which can cause major
performance problems.

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>, Matthias J. Sax <matthias@confluent.io>

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1315,4 +1450,22 @@ private Map<String, Object> structToMap(Struct struct) {
     }
+
+    // Manually insert a connector into config storage, updating the task configs, connector config, and root config
+    private void whiteboxAddConnector(String connectorName, Map<String, String> connectorConfig, List<Map<String, String>> taskConfigs) {
```
`whiteBoxAddConnector`

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+        verify(configLog).start();
+
+        ClusterConfigState configState = configStorage.snapshot();
+        expectRead(TARGET_STATE_KEYS.get(0), CONFIGS_SERIALIZED.get(0), TARGET_STATE_STARTED);
```
Maybe we can call `expectRead(Map<String, byte[]> serializedValues, Map<String, Struct> deserializedValues)` directly. For example:

```java
        expectRead(Collections.singletonMap(TARGET_STATE_KEYS.get(0), CONFIGS_SERIALIZED.get(0)),
            Collections.singletonMap(TARGET_STATE_KEYS.get(0), TARGET_STATE_STARTED));
```

BTW, we don't need to use `LinkedHashMap` in `expectRead`

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+        // Validate root config by listing all connectors and tasks
+        configState = configStorage.snapshot();
+        String connectorName = CONNECTOR_IDS.get(0);
+        assertEquals(Arrays.asList(connectorName), new ArrayList<>(configState.connectors()));
```
`assertEquals(Collections.singleton(connectorName), configState.connectors());`

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+        assertEquals(-1, configState.offset());
+
+        // Task configs should read to end, write to the log, read to end, write root.
+        doAnswer(expectReadToEnd(new LinkedHashMap<>())).when(configLog).readToEnd();
```
`Collections.emptyMap()`

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+                COMMIT_TASKS_CONFIG_KEYS.get(0), KafkaConfigBackingStore.CONNECTOR_TASKS_COMMIT_V0, CONFIGS_SERIALIZED.get(0),
+                "tasks", 0); // We have 0 tasks
+
+        List<Map<String, String>> taskConfigs = Collections.emptyList();
```
we don't need temp `taskConfigs` here

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+        String connectorName = CONNECTOR_IDS.get(0);
+        assertEquals(Arrays.asList(connectorName), new ArrayList<>(configState.connectors()));
+        assertEquals(Collections.emptyList(), configState.tasks(connectorName));
+        assertEquals(Collections.EMPTY_SET, configState.inconsistentConnectors());
```
`Collections.emptySet()`

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+        assertEquals(Collections.EMPTY_SET, configState.inconsistentConnectors());
+
+        configStorage.stop();
+        verify(configStorage).stop();
```
We should verify `configLog#stop`, as `configStorage` is just stopped before, right?
`verify(configLog).stop();`

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+        assertEquals(Collections.singleton(CONNECTOR_IDS.get(0)), configStorage.connectorTargetStates.keySet());
+        assertEquals(TargetState.STARTED, configState.targetState(CONNECTOR_IDS.get(0)));
+
+        LinkedHashMap<String, byte[]> serializedAfterStartup = new LinkedHashMap<>();
```
It seems we don't use the insert order, so maybe `HashMap` is enough .

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+        // Should see two connectors now, one paused and one stopped
+        configStorage.refresh(0, TimeUnit.SECONDS);
+        verify(configUpdateListener).onConnectorTargetStateChange(CONNECTOR_IDS.get(0));
+        configUpdateListener.onConnectorConfigRemove(CONNECTOR_IDS.get(0));
```
Why we need this? If you expect it should be called, please add `verify(configUpdateListener).onConnectorConfigRemove(CONNECTOR_IDS.get(0));` instead.

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,140 @@ public void testRestoreRestartRequestInconsistentState() {
+        assertEquals(TargetState.STOPPED, configState.targetState(CONNECTOR_IDS.get(1)));
+
+        configStorage.stop();
+        verify(configStorage).stop();
```
ditto

chia7712 reviewed (commented):
@chiacyu thanks for taking over this verbose issue!

brandboat committed:
MINOR: fix incorrect formatter package in streams quickstart (#15991)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

chiacyu committed:
Change from linkedhashmap to map

kirktrue committed:
KAFKA-15974: Enforce that event processing respects user-provided timeout (#15640)

The intention of the CompletableApplicationEvent is for a Consumer to enqueue the event and then block, waiting for it to complete. The application thread will block up to the amount of the timeout. This change introduces a consistent manner in which events are expired out by checking their timeout values.

The CompletableEventReaper is a new class that tracks CompletableEvents that are enqueued. Both the application thread and the network I/O thread maintain their own reaper instances. The application thread will track any CompletableBackgroundEvents that it receives and the network I/O thread will do the same with any CompletableApplicationEvents it receives. The application and network I/O threads will check their tracked events, and if any are expired, the reaper will invoke each event's CompletableFuture.completeExceptionally() method with a TimeoutException.

On closing the AsyncKafkaConsumer, both threads will invoke their respective reapers to cancel any unprocessed events in their queues. In this case, the reaper will invoke each event's CompletableFuture.completeExceptionally() method with a CancellationException instead of a TimeoutException to differentiate the two cases.

The overall design for the expiration mechanism is captured on the Apache wiki and the original issue (KAFKA-15848) has more background on the cause.

Note: this change only handles the event expiration and does not cover the network request expiration. That is handled in a follow-up Jira (KAFKA-16200) that builds atop this change.

This change also includes some minor refactoring of the EventProcessor and its implementations. This allows the event processor logic to focus on processing individual events rather than also the handling of batches of events.

Reviewers: Lianet Magrans <lianetmr@gmail.com>, Philip Nee <pnee@confluent.io>, Bruno Cadonna <cadonna@apache.org>

mimaison committed:
MINOR: Move Throttler to storage module (#16023)


Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mimaison committed:
KAFKA-12399: Deprecate KafkaLog4jAppender (#15985)


Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 5) Add rack information to ApplicationState (#15972)

This rack information is required to compute rack-aware assignments, which many of the current assigners do.

The internal ClientMetadata class was also edited to pass around this rack information.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 7) Simplify requirements for rack aware graphs (#16004)

Rack aware graphs don't actually need any topology information about the system, but rather require a simple ordered (not sorted) grouping of tasks.

This PR changes the internal constructors and some interface signatures of RackAwareGraphConstructor and its implementations to allow reuse by future components that may not have access to the actual subtopology information.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 8) Added TopicPartitionAssignmentInfo (#16024)

For task assignment purposes, the user needs to have a set of information available for each topic partition affecting the desired tasks.

This PR introduces a new interface for a read-only container class that allows all the important and relevant information to be found in one place.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

nicktelford committed:
KAFKA-15541: Add iterator-duration metrics (#16028)

Part of [KIP-989](https://cwiki.apache.org/confluence/x/9KCzDw).

This new `StateStore` metric tracks the average and maximum amount of
time between creating and closing Iterators.

Iterators with very high durations can indicate to users performance
problems that should be addressed.

If a store reports no data for these metrics, despite the user opening
Iterators on the store, it suggests those iterators are not being
closed, and have therefore leaked.

Reviewers: Matthias J. Sax <matthias@confluent.io>

jeffkbkim committed:
MINOR: fix flaky testRecordThreadIdleRatio (#15987)

DelayEventAccumulator should return immediately if there are no events in the queue. Also removed some unused fields inside EventProcessorThread.

Reviewers: Gaurav Narula <gaurav_narula2@apple.com>, Chia-Ping Tsai <chia7712@gmail.com>, David Jacot <djacot@confluent.io>

dongnuo123 committed:
KAFKA-16793; Heartbeat API for upgrading ConsumerGroup (#15988)

This patch implements the heartbeat api to the members that use the classic protocol in a ConsumerGroup.

Reviewers: Jeff Kim <jeff.kim@confluent.io>, David Jacot <djacot@confluent.io>

mimaison committed:
MINOR: Move parseCsvList to server-common (#16029)


Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

mimaison committed:
KAFKA-16825: Update netty/jetty/jackson/zstd dependencies (#16038)


Reviewers: Luke Chen <showuon@gmail.com>

viktorsomogyi committed:
KAFKA-15649: Handle directory failure timeout (#15697)

A broker that is unable to communicate with the controller will shut down
after the configurable log.dir.failure.timeout.ms.

The implementation adds a new event to the Kafka EventQueue. This event
is deferred by the configured timeout and will execute the shutdown
if the heartbeat communication containing the failed log dir is still
pending with the controller.

Reviewers: Igor Soarez <soarez@apple.com>

edoardocomar committed:
KAFKA-15905 Restarts of MirrorCheckpointTask should not permanently i… (#15910)

* KAFKA-15905 Restarts of MirrorCheckpointTask should not permanently interrupt offset translation

MirrorCheckpointTask reloads the last checkpoint at start,
OffsetSyncStore stores OffsetSyncs before reading till end.

If CheckpointTask cannot read checkpoints at startup,
use previous OffsetSyncStore load logic, with
warning log message about degraded offset translation.

Also addresses KAFKA-16622 Mirromaker2 first Checkpoint not emitted until
consumer group fully catches up once because the OffsetSyncStore store
is populated before reading to log end.

Co-Authored-By: Adrian Preston <prestona@uk.ibm.com>
Reviewers: Greg Harris <greg.harris@aiven.io>

frankvicky committed:
KAFKA-16804: Replace archivesBaseName with archivesName (#16016)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Greg Harris <greg.harris@aiven.io>

brandboat committed:
KAFKA-16795 Fix broken compatibility in kafka.tools.NoOpMessageFormatter, kafka.tools.DefaultMessageFormatter, and kafka.tools.LoggingMessageFormatter (#16020)

This commit allows users to apply the scala version Formatters, but users will receive the warning messages about deprecation. 

This compatibility support will be removed from 4.0.0
 
Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 10) Topic partition rack annotation simplified  (#16034)

This PR uses the new TaskTopicPartition structure to simplify the build
process for the ApplicationState, which is the input to the new
TaskAssignor#assign call.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

gharris1727 committed:
MINOR: Refactor Values class to fix checkstyle, add benchmark, optimize exceptions (#15469)

Signed-off-by: Greg Harris <greg.harris@aiven.io>
Reviewers: Mickael Maison <mickael.maison@gmail.com>

brandboat committed:
KAFKA-16828 RackAwareTaskAssignorTest failed (#16044)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

kagarwal06 committed:
KAFKA-16826: Integrate Native Docker Image with github actions (#16045)

This PR integrates the Native docker image with the existing github action jobs for the jvm docker image of AK.

The integration is done to the following actions:

docker_build_and_test.yml: Builds the docker image and runs sanity tests and CVE scan
docker_rc_release.yml: Builds the RC docker image for both amd and arm platform and pushes it to the dockerhub.
docker_promote.yml: Promotes the RC docker image to the released image tag

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Vedarth Sharma <142404391+VedarthConfluent@users.noreply.github.com>

jeffkbkim committed:
KAFKA-16626; Lazily convert subscribed topic names to topic ids (#15970)

This patch aims to remove the data structure that stores the conversion from topic names to topic ids which was taking time similar to the actual assignment computation. Instead, we reuse the already existing ConsumerGroupMember.subscribedTopicNames() and do the conversion to topic ids when the iterator is requested.

Reviewers: David Jacot <djacot@confluent.io>

tinaselenge committed:
MINOR: Refactor SSL/SASL admin integration tests to not use a custom authorizer (#15377)


Reviewers: Mickael Maison <mickael.maison@gmail.com>

lianetm committed:
KAFKA-16815: Handle FencedInstanceId in HB response (#16047)

Handle FencedInstanceIdException that a consumer may receive in the heartbeat response. This will be the case when a static consumer is removed from the group by and admin client, and another member joins with the same group.instance.id (allowed in). The first member will receive a FencedInstanceId on its next heartbeat. The expectation is that this should be handled as a fatal error.

There are no actual changes in logic with this PR, given that without being handled, the FencedInstanceId was being treated as an "unexpected error", which are all treated as fatal errors, so the outcome remains the same. But we're introducing this small change just for accuracy in the logic and the logs: FencedInstanceId is expected during heartbeat, a log line is shown describing the situation and why it happened (and it's treated as a fatal error, just like it was before this PR).

This PR also improves the test to ensure that the error propagated to the app thread matches the one received in the HB.

Reviewers: Andrew Schofield <aschofield@confluent.io>, David Jacot <djacot@confluent.io>

KrishVora01 committed:
KAFKA-16373: KIP-1028:  Adding code to support Apache Kafka Docker Official Images (#16027)

This PR aims to add JVM based Docker Official Image for Apache Kafka as per the following KIP - https://cwiki.apache.org/confluence/display/KAFKA/KIP-1028%3A+Docker+Official+Image+for+Apache+Kafka

This PR adds the following functionalities:
Introduces support for Apache Kafka Docker Official Images via:

GitHub Workflows:

- Workflow to prepare static source files for Docker images
- Workflow to build and test Docker official images
- Scripts to prepare source files and perform Docker image builds and tests

A new directory for Docker official images, named docker/docker_official_images. This is the new directory to house all Docker Official Image assets.

Co-authored-by: Vedarth Sharma <vesharma@confluent.io>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Vedarth Sharma <vesharma@confluent.io>

edoardocomar committed:
MINOR: mention KAFKA-15905 in docs "Notable changes in 3.7.1" (#16070)

* MINOR: mention KAFKA-15905 in docs "Notable changes in 3.7.1/3.8.0"

Co-Authored-By: Adrian Preston <prestona@uk.ibm.com>

jeffkbkim committed:
KAFKA-16831: CoordinatorRuntime should initialize MemoryRecordsBuilder with max batch size write limit (#16059)

CoordinatorRuntime should initialize MemoryRecordsBuilder with max batch size write limit. Otherwise, we default the write limit to the min buffer size of 16384 for the write limit. This causes the coordinator to threw RecordTooLargeException even when it's under the 1MB max batch size limit.

Reviewers: David Jacot <djacot@confluent.io>

rreddy-22 committed:
KAFKA-16625; Reverse lookup map from topic partitions to members (#15974)

This patch speeds up the computation of the unassigned partitions by exposing the inverted target assignment. It allows the assignor to check whether a partition is assigned or not.

Reviewers: Jeff Kim <jeff.kim@confluent.io>, David Jacot <djacot@confluent.io>

nicktelford committed:
KAFKA-15541: Use LongAdder instead of AtomicInteger (#16076)

`LongAdder` performs better than `AtomicInteger` when under contention
from many threads. Since it's possible that many Interactive Query
threads could create a large number of `KeyValueIterator`s, we don't
want contention on a metric to be a performance bottleneck.

The trade-off is memory, as `LongAdder` uses more memory to space out
independent counters across different cache lines. In practice, I don't
expect this to cause too many problems, as we're only constructing 1
per-store.

Reviewers: Matthias J. Sax <matthias@confluent.io>

kamalcph committed:
KAFKA-16452: Don't throw OOORE when converting the offset to metadata (#15825)

Don't throw OFFSET_OUT_OF_RANGE error when converting the offset to metadata, and next time the leader should increment the high watermark by itself after receiving fetch requests from followers. This can happen when checkpoint files are missing and being elected as a leader. 

Reviewers: Luke Chen <showuon@gmail.com>, Jun Rao <junrao@apache.org>

dajac committed:
KAFKA-16371; fix lingering pending commit when handling OFFSET_METADATA_TOO_LARGE (#16072)

This patch was initially created in #15536.

When there is a commit for multiple topic partitions and some, but not all, exceed the offset metadata limit, the pending commit is not properly cleaned up leading to UNSTABLE_OFFSET_COMMIT errors when trying to fetch the offsets with read_committed. This change makes it so the invalid commits are not added to the pendingOffsetCommits set.

Co-authored-by: Kyle Phelps <kyle.phelps@datadoghq.com>

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Justine Olshan <jolshan@confluent.io>

cmccabe committed:
MINOR: fix typo in KAFKA-16515

gharris1727 committed:
MINOR: Disable JDK 11 and 17 tests on PRs (#16051)

Signed-off-by: Greg Harris <greg.harris@aiven.io>
Reviewers: Justine Olshan <jolshan@confluent.io>, David Arthur <mumrah@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Luke Chen <showuon@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

showuon committed:
KAFKA-16709: abortAndPauseCleaning only when future log is not existed (#15951)

When doing alter replica logDirs, we'll create a future log and pause log cleaning for the partition( here). And this log cleaning pausing will resume after alter replica logDirs completes (here). And when in the resuming log cleaning, we'll decrement 1 for the LogCleaningPaused count. Once the count reached 0, the cleaning pause is really resuming. (here). For more explanation about the logCleaningPaused state can check here.

But, there's still one factor that could increase the LogCleaningPaused count: leadership change (here). When there's a leadership change, we'll check if there's a future log in this partition, if so, we'll create future log and pauseCleaning (LogCleaningPaused count + 1). So, if during the alter replica logDirs:

1. alter replica logDirs for tp0 triggered (LogCleaningPaused count = 1)
2. tp0 leadership changed (LogCleaningPaused count = 2)
3. alter replica logDirs completes, resuming logCleaning (LogCleaningPaused count = 1)
4. LogCleaning keeps paused because the count is always >  0

This PR fixes this issue by only abortAndPauseCleaning when future log is not existed. We did the same check in alterReplicaLogDirs. So this change can make sure there's only 1 abortAndPauseCleaning for either abortAndPauseCleaning or maybeAddLogDirFetchers. Tests also added.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Igor Soarez <soarez@apple.com>

m1a2st committed:
KAFKA-16805 Stop using a ClosureBackedAction to configure Spotbugs reports (#16081)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

sjhajharia committed:
MINOR: Fix the config name in ProducerFailureHandlingTest (#16099)

When moving from KafkaConfig.ReplicaFetchMaxBytesProp we used ReplicationConfigs.REPLICA_LAG_TIME_MAX_MS_CONFIG instead of ReplicationConfigs.REPLICA_FETCH_MAX_BYTES_CONFIG. This PR patches the same.

Reviewers: Omnia Ibrahim <o.g.h.ibrahim@gmail.com>, Manikumar Reddy <manikumar.reddy@gmail.com>

OmniaGM committed:
KAFKA-15853: Move general configs out of KafkaConfig (#16040)


Reviewers: Mickael Maison <mickael.maison@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

showuon committed:
KAFKA-16711: Make sure to update highestOffsetInRemoteStorage after log dir change (#15947)

Reviewers: Kamal Chandraprakash<kamal.chandraprakash@gmail.com>, Satish Duggana <satishd@apache.org>

FrankYang0529 committed:
KAFKA-16796 Introduce new org.apache.kafka.tools.api.Decoder to replace kafka.serializer.Decoder (#16064)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

fred-ro committed:
KAFKA-16507 Add KeyDeserializationException and ValueDeserializationException with record content (#15691)

Implements KIP-1036.

Add raw ConsumerRecord data to RecordDeserialisationException to make DLQ implementation easier.

Reviewers: Kirk True <ktrue@confluent.io>, Andrew Schofield <aschofield@confluent.io>, Matthias J. Sax <matthias@confluent.io>

nicktelford committed:
KAFKA-15541: Add oldest-iterator-open-since-ms metric (#16041)

Part of [KIP-989](https://cwiki.apache.org/confluence/x/9KCzDw).

This new `StateStore` metric tracks the timestamp that the oldest
surviving Iterator was created.

This timestamp should continue to climb, and closely track the current
time, as old iterators are closed and new ones created. If the timestamp
remains very low (i.e. old), that suggests an Iterator has leaked, which
should enable users to isolate the affected store.

It will report no data when there are no currently open Iterators.

Reviewers: Matthias J. Sax <matthias@confluent.io>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 11) Implemented StickyTaskAssignor (#16052)

This PR implements the StickyTaskAssignor with the new KIP 924 API.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 12) Wiring in new assignment configs and logic (#16074)

This PR creates the new public config of KIP-924 in StreamsConfig and uses it to instantiate user-created TaskAssignors. If such a TaskAssignor is found and successfully created we then use that assignor to perform the task assignment, otherwise we revert back to the pre KIP-924 world with the internal task assignors.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>, Almog Gavra <almog@responsive.dev>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 13) AssignmentError calculation added (#16114)

This PR adds the post-processing of the TaskAssignment to figure out if the new assignment is valid, and return an AssignmentError otherwise.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

ableegoldman committed:
HOTFIX: remove unnecessary list creation (#16117)

Removing a redundant list declaration in the new StickyTaskAssignor implementation

Reviewers: Antoine Pourchet <antoine@responsive.dev>

dongnuo123 committed:
KAFKA-16832; LeaveGroup API for upgrading ConsumerGroup (#16057)

This patch implements the LeaveGroup API to the consumer groups that are in the mixed mode.

Reviewers: Jeff Kim <jeff.kim@confluent.io>, David Jacot <djacot@confluent.io>

showuon committed:
KAFKA-16399: Add JBOD support in tiered storage (#15690)

After JBOD is supported in KRaft, we should also enable JBOD support in tiered storage. Unit tests and Integration tests are also added.

Reviewers: Satish Duggana <satishd@apache.org>, Kamal Chandraprakash <kamal.chandraprakash@gmail.com>, Igor Soarez <soarez@apple.com>, Mickael Maison <mickael.maison@gmail.com>

raminqaf committed:
KAFKA-16362: Fix type-unsafety in KStreamKStreamJoin caused by isLeftSide (#15601)

The introduced changes provide a cleaner definition of the join side in KStreamKStreamJoin. Before, this was done by using a Boolean flag, which led to returning a raw LeftOrRightValue without generic arguments because the generic type arguments depended on the boolean input.

Reviewers: Greg Harris <greg.harris@aiven.io>, Bruno Cadonna <cadonna@apache.org>

loicgreffier committed:
KAFKA-16448: Add ProcessingExceptionHandler interface and implementations (#16090)

This PR is part of KAFKA-16448 which aims to bring a ProcessingExceptionHandler to Kafka Streams in order to deal with exceptions that occur during processing.

This PR brings ProcessingExceptionHandler interface and default implementations.

Co-authored-by: Dabz <d.gasparina@gmail.com>
Co-authored-by: sebastienviale <sebastien.viale@michelin.com>

Reviewer: Bruno Cadonna <cadonna@apache.org>

gongxuanzhang committed:
KAFKA-16705 the flag "started" of RaftClusterInstance is false even though the cluster is started (#15946)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

AndrewJSchofield committed:
KAFKA-16722: Introduce ConsumerGroupPartitionAssignor interface (#15998)

KIP-932 introduces share groups to go alongside consumer groups. Both kinds of group use server-side assignors but it is unlikely that a single assignor class would be suitable for both. As a result, the KIP introduces specific interfaces for consumer group and share group partition assignors.

This PR introduces only the consumer group interface, `o.a.k.coordinator.group.assignor.ConsumerGroupPartitionAssignor`. The share group interface will come in a later release. The existing implementations of the general `PartitionAssignor` interface have been changed to implement `ConsumerGroupPartitionAssignor` instead and all other code changes are just propagating the change throughout the codebase.

Note that the code in the group coordinator that actually calculates assignments uses the general `PartitionAssignor` interface so that it can be used with both kinds of group, even though the assignors themselves are specific.

Reviewers: Apoorv Mittal <amittal@confluent.io>, David Jacot <djacot@confluent.io>

gongxuanzhang committed:
KAFKA-16771 First log directory printed twice when formatting storage (#16010)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

emitskevich-blp committed:
MINOR: Fix rate metric spikes (#15889)

Rate reports value in the form of sumOrCount/monitoredWindowSize. It has a bug in monitoredWindowSize calculation, which leads to spikes in result values.

Reviewers: Jun Rao <junrao@gmail.com>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 14) Callback to TaskAssignor::onAssignmentComputed (#16123)

This PR adds the logic and wiring necessary to make the callback to
TaskAssignor::onAssignmentComputed with the necessary parameters.

We also fixed some log statements in the actual assignment error
computation, as well as modified the ApplicationState::allTasks method
to return a Map instead of a Set of TaskInfos.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

mimaison committed:
MINOR: Delete KafkaSecurityConfigs class (#16113)

Reviewers: Omnia Ibrahim <o.g.h.ibrahim@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 9) TaskAssignmentUtils implementation of optimizeRackAwareActiveTasks (#16033)

This PR implements the rack aware optimization of active tasks that can be used by the assignors themselves. It takes in the full output of the assignment and tries to reorganize it so as to minimize cross-rack traffic.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

jolshan committed:
KAFKA-16308 [1/N]: Create FeatureVersion interface and add `--feature` flag and handling to StorageTool (#15685)

As part of KIP-1022, I have created an interface for all the new features to be used when parsing the command line arguments, doing validations, getting default versions, etc.

I've also added the --feature flag to the storage tool to show how it will be used.

Created a TestFeatureVersion to show an implementation of the interface (besides MetadataVersion which is unique) and added tests using this new test feature.

I will add the unstable config and tests in a followup.

Reviewers: David Mao <dmao@confluent.io>, David Jacot <djacot@confluent.io>, Artem Livshits <alivshits@confluent.io>, Jun Rao <junrao@apache.org>

CalvinLiu7947 committed:
Improve producer ID expiration performance (#16075)

Skip using stream when expiring the producer ID. This can improve the performance significantly when the count is high.
Before

Benchmark                                        (numProducerIds)  Mode  Cnt      Score       Error  Units
ProducerStateManagerBench.testDeleteExpiringIds             10000  avgt    3    101.253 ±    28.031  us/op
ProducerStateManagerBench.testDeleteExpiringIds            100000  avgt    3   2297.219 ±  1690.486  us/op
ProducerStateManagerBench.testDeleteExpiringIds           1000000  avgt    3  30688.865 ± 16348.768  us/op
After

Benchmark                                        (numProducerIds)  Mode  Cnt     Score     Error  Units
ProducerStateManagerBench.testDeleteExpiringIds             10000  avgt    3    39.122 ±   1.151  us/op
ProducerStateManagerBench.testDeleteExpiringIds            100000  avgt    3   464.363 ±  98.857  us/op
ProducerStateManagerBench.testDeleteExpiringIds           1000000  avgt    3  5731.169 ± 674.380  us/op
Also, made a change to the JMH testing which excludes the producer ID populating from the testing.

Reviewers: Artem Livshits <alivshits@confluent.io>, Justine Olshan <jolshan@confluent.io>

muralibasani committed:
KAFKA-16790: Update RemoteLogManager configuration in broker server (#16005)

n BrokerServer.scala, brokerMetadataPublishers are configured and when there are metadata updates remoteLogManager is not configured by then.
Ex : remoteLogManager.foreach(rlm => rlm.onLeadershipChange(partitionsBecomeLeader.asJava, partitionsBecomeFollower.asJava, topicIds)) in ReplicaManager is invoked after publishers are instantiated, and here rlm has relevant managers configured.

This change makes sure rlm is configured before the brokerMetadataPublishers initialization.

Reviewers: Luke Chen <showuon@gmail.com>, Nikhil Ramakrishnan <nikrmk@amazon.com>

dengziming committed:
Minor: Fix VoterSetHistoryTest.testAddAt (#16104)

Reviewers: Luke Chen <showuon@gmail.com>

dajac committed:
MINOR: Prevent consumer protocol to be used in ZK mode (#16121)

This patch disallows enabling the new consumer rebalance protocol in ZK mode.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Justine Olshan <jolshan@confluent.io>

loicgreffier committed:
KAFKA-16448: Add ProcessingExceptionHandler in Streams configuration (#16092)

This PR is part of KAFKA-16448 which aims to bring a ProcessingExceptionHandler to Kafka Streams in order to deal with exceptions that occur during processing.

This PR brings ProcessingExceptionHandler in Streams configuration.

Co-authored-by: Dabz <d.gasparina@gmail.com>
Co-authored-by: sebastienviale <sebastien.viale@michelin.com>

Reviewer: Bruno Cadonna <cadonna@apache.org>

mimaison committed:
MINOR: Remove KafkaConfig dependency in KafkaRequestHandler (#16108)


Reviewers: Luke Chen <showuon@gmail.com>, Apoorv Mittal <amittal@confluent.io>

FrankYang0529 committed:
MINOR: migrate ListConsumerGroupTest to use ClusterTestExtensions (#15821)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

m1a2st committed:
KAFKA-16598 Mirgrate `ResetConsumerGroupOffsetTest` to new test infra (#15779)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

chiacyu committed:
Applied review comments

fanyang committed:
KAFKA-16844: Add ByteBuffer support for Connect ByteArrayConverter (#16101)

Reviewers: Chris Egerton <chrise@aiven.io>

cadonna committed:
Revert "KAFKA-16448: Add ProcessingExceptionHandler in Streams configuration (#16092)" (#16141)

This reverts commit 3f70c46874e1dd9591443392f51ff1efc9fdc40e.

Reviewer: Lucas Brutschy <lbrutschy@confluent.io>

abhijeetk88 committed:
KAFKA-15265: Add Remote Log Manager quota manager (#15625)

Added the implementation of the quota manager that will be used to throttle copy and fetch requests from the remote storage. Reference KIP-956

Reviewers: Luke Chen <showuon@gmail.com>, Kamal Chandraprakash <kchandraprakash@uber.com>, Jun Rao <junrao@gmail.com>

kagarwal06 committed:
KAFKA-16827: Integrate kafka native-image with system tests (#16046)

This PR does following things

System tests should bring up Kafka broker in the native mode
System tests should run on Kafka broker in native mode
Extract out native build command so that it can be reused.
Allow system tests to run on Native Kafka broker using Docker mechanism

To run system tests by bringing up Kafka in native mode:
Pass kafka_mode as native in the ducktape globals:--globals '{\"kafka_mode\":\"native\"}'

Running system tests by bringing up kafka in native mode via docker mechanism
_DUCKTAPE_OPTIONS="--globals '{\"kafka_mode\":\"native\"}'" TC_PATHS="tests/kafkatest/tests/"  bash tests/docker/run_tests.sh

To only bring up ducker nodes to cater native kafka
bash tests/docker/ducker-ak up -m native

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

dongnuo123 committed:
MINOR: Make public the consumer group migration policy config

This patch exposes the group coordinator config `CONSUMER_GROUP_MIGRATION_POLICY_CONFIG`.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, David Jacot <djacot@confluent.io>

muralibasani committed:
KAFKA-16802: Move java versions inside java block to resolve deprecation (#16135)

Reviewers: Greg Harris <greg.harris@aiven.io>

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1184,6 +1185,141 @@ public void testRestoreRestartRequestInconsistentState() {
+        configStorage.putTaskConfigs("connector1", Collections.emptyList());
+
+        // As soon as root is rewritten, we should see a callback notifying us that we reconfigured some tasks
+        configUpdateListener.onTaskConfigUpdate(Collections.emptyList());
```
We need to use `verify` to make sure this method is called as expected.

chia7712 replied:
Please see following example:

```java
    @Test
    public void testPutTaskConfigsZeroTasks() throws Exception {
        configStorage.setupAndCreateKafkaBasedLog(TOPIC, config);
        verifyConfigure();
        configStorage.start();
        verify(configLog).start();

        // Records to be read by consumer as it reads to the end of the log
        doAnswer(expectReadToEnd(new LinkedHashMap<>())).
            doAnswer(expectReadToEnd(Collections.singletonMap(COMMIT_TASKS_CONFIG_KEYS.get(0), CONFIGS_SERIALIZED.get(0))))
            .when(configLog).readToEnd();

        expectConvertWriteRead(
            COMMIT_TASKS_CONFIG_KEYS.get(0), KafkaConfigBackingStore.CONNECTOR_TASKS_COMMIT_V0, CONFIGS_SERIALIZED.get(0),
            "tasks", 0); // We have 0 tasks

        // Bootstrap as if we had already added the connector, but no tasks had been added yet
        addConnector(CONNECTOR_IDS.get(0), SAMPLE_CONFIGS.get(0), Collections.emptyList());


        // Null before writing
        ClusterConfigState configState = configStorage.snapshot();
        assertEquals(-1, configState.offset());

        // Writing task configs should block until all the writes have been performed and the root record update
        // has completed
        List<Map<String, String>> taskConfigs = Collections.emptyList();
        configStorage.putTaskConfigs("connector1", taskConfigs);

        // Validate root config by listing all connectors and tasks
        configState = configStorage.snapshot();
        assertEquals(1, configState.offset());
        String connectorName = CONNECTOR_IDS.get(0);
        assertEquals(Collections.singletonList(connectorName), new ArrayList<>(configState.connectors()));
        assertEquals(Collections.emptyList(), configState.tasks(connectorName));
        assertEquals(Collections.EMPTY_SET, configState.inconsistentConnectors());

        // As soon as root is rewritten, we should see a callback notifying us that we reconfigured some tasks
        verify(configUpdateListener).onTaskConfigUpdate(Collections.emptyList());

        configStorage.stop();
        verify(configLog).stop();
    }
```

chiacyu replied:
Thanks for the example, would apply.

dajac committed:
MINOR: Enable transaction verification with new group coordinator in TransactionsTest (#16139)

While working on https://github.com/apache/kafka/pull/16120, I noticed that the transaction verification feature is disabled in `TransactionsTest` when the new group coordinator is enabled. We did this initially because the feature was not available in the new group coordinator but we fixed it a long time ago. We can enable it now.

Reviewers: Justine Olshan <jolshan@confluent.io>

ahmedryasser committed:
MINOR: Adds a test case to test that an exception is thrown in invalid ports (#16112)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

sjhajharia committed:
MINOR: Code Cleanup - Connect Module (#16066)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

ahuang98 committed:
KAFKA-16833 Fixing PartitionInfo and Cluster equals and hashCode (#16062)

Reviewers: Andrew Schofield <aschofield@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

jolshan committed:
KAFKA-16308 [2/N]: Allow unstable feature versions and rename unstable metadata config (#16130)

As per KIP-1022, we will rename the unstable metadata versions enabled config to support all feature versions.

Features is also updated to return latest production and latest testing versions of each feature.

A feature is production ready when the corresponding metadata version (bootstrapMetadataVersion) is production ready.

Adds tests for the feature usage of the unstableFeatureVersionsEnabled config

Reviewers: David Jacot <djacot@confluent.io>, Jun Rao <junrao@gmail.com>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 15) Implement #defaultStandbyTaskAssignment and finish rack-aware standby optimization (#16129)

This fills in the implementation details of the standby task assignment utility functions within TaskAssignmentUtils.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

chia7712 committed:
KAFKA-16866 RemoteLogManagerTest.testCopyQuotaManagerConfig failing (#16146)

Reviewers: Justine Olshan <jolshan@confluent.io>, Satish Duggana <satishd@apache.org>

cadonna committed:
Revert "KAFKA-16448: Add ProcessingExceptionHandler interface and implementations (#16090)" (#16142)

This reverts commit 8d11d9579503426edfaeae791ec4bb212da37ad2.

We decided to not release KIP-1033 with AK 3.8

Reviewer: Lucas Brutschy <lbrutschy@confluent.io>

brandboat committed:
MINOR: Add more unit tests to LogSegments (#16085)

add more unit tests to LogSegments and do some small refactor in LogSegments.java

Reviewers: Luke Chen <showuon@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

lianetm committed:
KAFKA-16766: offset fetch timeout exception in new consumer consistent with legacy (#16125)

* Timeout exception fetching offsets

* Tests

frankvicky committed:
KAFKA-16824: Utils.getHost and Utils.getPort do not catch a lot of invalid host and ports. (#16048)

Modify regex of HOST_PORT_PATTERN to prevent malformed hosts and ports.

Reviewers: Luke Chen <showuon@gmail.com>

m1a2st committed:
KAFKA-16629 Add broker-related tests to ConfigCommandIntegrationTest (#15840)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java:
```diff
@@ -1315,4 +1451,22 @@ private Map<String, Object> structToMap(Struct struct) {
     }
+
+    // Manually insert a connector into config storage, updating the task configs, connector config, and root config
+    private void whiteBoxAddConnector(String connectorName, Map<String, String> connectorConfig, List<Map<String, String>> taskConfigs) {
```
we don't use `whiteBox` now, so that can be simplified:
```java
    private void addConnector(String connectorName, Map<String, String> connectorConfig, List<Map<String, String>> taskConfigs) {
        for (int i = 0; i < taskConfigs.size(); i++)
            configStorage.taskConfigs.put(new ConnectorTaskId(connectorName, i), taskConfigs.get(i));
        configStorage.connectorConfigs.put(connectorName, connectorConfig);
        configStorage.connectorTaskCounts.put(connectorName, taskConfigs.size());
    }
```

chiacyu replied:
I would open a new pr for it.
Please check: https://github.com/apache/kafka/pull/16164, thanks!

chiacyu committed:
Rebased truck and applied review comments

chiacyu committed:
Rebased truck and applied review comments
